doc/wmgenmenu.1
.TH wmgenmenu 1 "15 August 2011"
.SH "NAME"
\fBwmgenmenu\fR \- Writes a menu structure usable as
~/GNUstep/Defaults/WMRootMenu to stdout
.PP
.SH "SYNOPSIS"
.B wmgenmenu
.RI [ options ]
.PP
.SH "DESCRIPTION"
\fBwmgenmenu\fR writes a menu structure usable as
~/GNUstep/Defaults/WMRootMenu to stdout.

The output format is a \fIPropList\fP.
.PP
.SH "OPTIONS"
.TP
.B \-\-help
print a help message with the list of options
.TP
.B \-\-version
print the version of Window Maker from which the program comes
.PP
.SH "BUGS"
The program is using a built-in structure and list of applications, which means it cannot discover
new programs, but it filters the programs that cannot be found in the \fBPATH\fP.
.SH "AUTHOR"
wmgenmenu is a part of Window Maker. It was written by Carlos R. Mafra and
Tamas TEVESZ.
.PP
This manpage was written by Rodolfo García Peñas (kix) for the
Debian GNU/Linux system (but may be used by others).


NEWS

NEWS for veteran Window Maker users
-----------------------------------

-- 0.95.7

Window snapping
---------------

You can now "snap" a window, i.e., maximize it to a side or corner of the
screen, by dragging it to that side or corner.  It is enabled by setting
"WindowSnapping = YES" in ~/GNUstep/Defaults/WindowMaker or selecting "Enable
window snapping" under "Expert User Preferences" in WPrefs.app.

Note that window snapping is automatically disabled if "Switch workspaces while
dragging windows" is selected under "Workspace Preferences" in WPrefs.app, or
if "DontLinkWorkspaces = NO" in  ~/GNUstep/Defaults/WindowMaker, as this feature
also involves dragging a window to one side of the screen.


Dragging maximized windows
--------------------------
You can now control the behavior when a maximized window is dragged by setting
the "DragMaximizedWindow" option in ~/GNUstep/Defaults/WindowMaker or by
selecting an option from the "When dragging a maximized window..." pop-up button
under "Window Handling Preferences" in WPrefs.app.

There are four choices:
* "Move" ("...change position (normal behavior)" in WPrefs.app) is the default
  and traditional behavior.  A maximized window is moved when dragged and
  remains maximized, i.e., it keeps its maximized geometry and can later be
  unmaximized.
* "RestoreGeometry" ("...restore unmaximized geometry") is the behavior standard
  in desktop environments like GNOME, Cinnamon, and Unity.  A maximized window
  is moved when dragged and is completely unmaximized, i.e., its unmaximized
  geometry is restored.
* "Unmaximize" ("...consider the window unmaximized") causes a maximized window
  to be moved when dragged and remains partially maximized, i.e., it keeps its
  maximized geometry, but is consider to be unmaximized.  In particular, it can
  be immediately re-maximized.
* "NoMove" ("...do not move the window") prevents a maximized window from being
  moved when dragged.

Note that, to accomodate this option in the "Window Handling Preferences" tab in
WPrefs.app, the option to "Open dialogs in the same workspace as their owners"
(which sets the "OpenTransientOnOwnerWorkspace" option from
~/GNUstep/Defaults/WindowMaker) has been moved to "Expert User Preferences".


Mini-Previews instead of Apercus
--------------------------------

Since the original name was not really clear because it is a French word that
is rarely used by British people, it was decided to change it to the more usual
Mini-Preview name. The setting is configurable with WPrefs in the Icon
Preferences tab, the size is now expressed in pixels directly.


--- 0.95.6

More image format supported
---------------------------

In addition to a more complete Netpbm image formats support, WindowMaker
can now load WebP images. It can also make use of the ImageMagick library
to support a lot more formats, like SVG, BMP, TGA...


Mini-window apercu
------------------

A small preview of window contents can be enabled from WPrefs,
in Miscellaneous Ergonomic Preferences, check miniwindow apercus.
Apercu size can be configured using the ApercuSize variable with

$ wdwrite WindowMaker ApercuSize 4

in multiples of the icon size (in this case the apercu size will be four
times the icon size).

The default size is 2 (twice the icon size).


Support for up to 9-buttons mouse
---------------------------------

The action for the newly supported buttons can be configured from WPrefs.


wmiv, an image viewer application
---------------------------------

wmiv is a quick image viewer using wrlib to be run from the command line.


--- 0.95.5

Support for generated menus in proplist format
----------------------------------------------

The root menu now supports a OPEN_PLMENU option to construct a submenu from
the output of a command which is proplist format. This can be used e.g. in
conjunction with wmmenugen like:

(
    "Generated PL Submenu", OPEN_PLMENU,
    "|| find /usr/share/applications -type f -name '*desktop' | xargs wmmenugen -parser:xdg"
)


New window placements
---------------------

Now it is possible to maximize windows to the top/bottom halves
of the screen and also to the corners (top/bottom + left/right).

The keyboard shortcuts can be configured with WPrefs.


Options to configure window/menu borders
----------------------------------------

You can now configure the width and color of window and menu borders.
For example, the default settings could be configured as follows:

$ wdwrite WindowMaker FrameBorderWidth 1
$ wdwrite WindowMaker FrameBorderColor black
$ wdwrite WindowMaker FrameSelectedBorderColor white


Keyboard shortcuts to move windows between workspaces
-----------------------------------------------------

You can now bind keyboard shortcuts - or use the window dropdown menus -
to move windows to other workspaces.  You can either move a window directly to
a particular workspace or to the "next" or "previous" workspace.

The new shortcuts can be configured in WPrefs.


Native support for Drawers
--------------------------

WindowMaker now supports drawers in the dock natively. You can add a dock by
right-clicking on a docked appicon and select "Add a drawer". You can now drag
appicons to this drawer.
You can customize how you want your drawers to auto-expand/collapse and
auto-raise/lower, or you can also completely disable them.


Improved switch panel functionality
-----------------------------------

The switch panel can be used to switch between windows of the same WM_CLASS,
for example between all open xterms.  If the switch panel is opened as normal
with either the "FocusNextKey" or "FocusPrevKey" shortcut, you can switch to
the next (or previous) window of the same type as the focused window with the
"GroupNextKey" or "GroupPrevKey" shortcut.  If the switch panel is opened with
the "GroupNextKey" or "GroupPrevKey" shortcut, it will show only windows of
the same type as the window which was focused at the time the panel was opened,

util/wmgenmenu.c
/* Copyright (C) 2010 Carlos R. Mafra */

#ifdef __GLIBC__
#define _GNU_SOURCE		/* getopt_long */
#endif

#include "config.h"

#include <ctype.h>
#include <getopt.h>
#include <limits.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#ifdef HAVE_STDNORETURN
#include <stdnoreturn.h>
#endif

#include <WINGs/WUtil.h>

#include "../src/wconfig.h"

#define MAX_NR_APPS 128 /* Maximum number of entries in each apps list */
#define MAX_WMS 10      /* Maximum number of other window managers to check */

#include "wmgenmenu.h"

static void find_and_write(const char *group, char *list[][2], int this_is_terminals);
static void other_window_managers(void);
static void print_help(int print_usage, int exitval);

extern char *__progname;

char *path, *terminal = NULL;

WMPropList *RMenu, *L1Menu, *L2Menu, *L3Menu, *L4Menu;

int main(int argc, char *argv[])
{
	char *t;
	int ch;
	char *tmp, *theme_paths, *style_paths, *icon_paths;

	tmp = wstrconcat("-noext ", PKGDATADIR);
	theme_paths = wstrconcat(tmp, "/Themes $HOME/GNUstep/Library/WindowMaker/Themes WITH setstyle");
	style_paths = wstrconcat(tmp, "/Styles $HOME/GNUstep/Library/WindowMaker/Styles WITH setstyle");
	icon_paths = wstrconcat(tmp, "/IconSets $HOME/GNUstep/Library/WindowMaker/IconSets WITH seticons");

	struct option longopts[] = {
		{ "version",		no_argument,	NULL,	'v' },
		{ "help",		no_argument,	NULL,	'h' },
		{ NULL,			0,		NULL,	0 }
	};

	while ((ch = getopt_long(argc, argv, "hv", longopts, NULL)) != -1)
		switch (ch) {
		case 'v':
			printf("%s (Window Maker %s)\n", __progname, VERSION);
			return 0;
			/* NOTREACHED */
		case 'h':
			print_help(1, 0);
			/* NOTREACHED */
		default:
			print_help(0, 1);
			/* NOTREACHED */
		}

	argc -= optind;
	argv += optind;

	if (argc != 0)
		print_help(0, 1);

	path = getenv("PATH");
	setlocale(LC_ALL, "");

#if defined(HAVE_LIBINTL_H) && defined(I18N)
	if (getenv("NLSPATH"))
		bindtextdomain("wmgenmenu", getenv("NLSPATH"));
	else
		bindtextdomain("wmgenmenu", LOCALEDIR);

	bind_textdomain_codeset("wmgenmenu", "UTF-8");
	textdomain("wmgenmenu");
#endif

	/*
	 * The menu generated is a five-level hierarchy, of which the
	 * top level (RMenu) is only used to hold the others (a single
	 * PLString, which will be the title of the root menu)
	 *
	 * RMenu                Window Maker
	 *   L1Menu               Applications
	 *     L2Menu               Terminals
	 *       L3Menu               XTerm
	 *       L3Menu               RXVT
	 *     L2Menu               Internet
	 *       L3Menu               Firefox
	 *     L2Menu               E-mail
	 *   L1Menu               Appearance
	 *     L2Menu               Themes
	 *     L2Menu               Background
	 *       L3Menu               Solid
	 *         L4Menu               Indigo
	 *   L1Menu               Configure Window Maker
	 *
	 */

	/* Root */
	RMenu = WMCreatePLArray(WMCreatePLString("Window Maker"), NULL);

	/* Root -> Applications */
	L1Menu = WMCreatePLArray(WMCreatePLString(_("Applications")), NULL);

	/* Root -> Applications -> <category> */
	find_and_write(_("Terminals"), Terminals, 1);	/* always keep terminals the top item */
	find_and_write(_("Internet"), Internet, 0);
	find_and_write(_("Email"), Email, 0);
	find_and_write(_("Mathematics"), Mathematics, 0);
	find_and_write(_("File Managers"), File_managers, 0);
	find_and_write(_("Graphics"), Graphics, 0);
	find_and_write(_("Multimedia"), Multimedia, 0);
	find_and_write(_("Editors"), Editors, 0);
	find_and_write(_("Development"), Development, 0);
	find_and_write("Window Maker", WindowMaker, 0);
	find_and_write(_("Office"), Office, 0);
	find_and_write(_("Astronomy"), Astronomy, 0);
	find_and_write(_("Sound"), Sound, 0);
	find_and_write(_("Comics"), Comics, 0);
	find_and_write(_("Viewers"), Viewers, 0);
	find_and_write(_("Utilities"), Utilities, 0);
	find_and_write(_("System"), System, 0);
	find_and_write(_("Video"), Video, 0);
	find_and_write(_("Chat and Talk"), Chat, 0);
	find_and_write(_("P2P Network"), P2P, 0);
	find_and_write(_("Games"), Games, 0);
	find_and_write("OpenSUSE", OpenSUSE, 0);
	find_and_write("Mandriva", Mandriva, 0);

	WMAddToPLArray(RMenu, L1Menu);

	/* Root -> `Run' dialog */
	L1Menu = WMCreatePLArray(
		WMCreatePLString(_("Run...")),
		WMCreatePLString("SHEXEC"),
		WMCreatePLString(_("%A(Run, Type command:)")),
		NULL
	);
	WMAddToPLArray(RMenu, L1Menu);

	/* Root -> Appearance */
	L1Menu = WMCreatePLArray(WMCreatePLString(_("Appearance")), NULL);

	/* Root -> Appearance -> Themes */
	L2Menu = WMCreatePLArray(
		WMCreatePLString(_("Themes")),
		WMCreatePLString("OPEN_MENU"),
		WMCreatePLString(theme_paths),
		NULL
	);
	WMAddToPLArray(L1Menu, L2Menu);

	/* Root -> Appearance -> Styles */
	L2Menu = WMCreatePLArray(
		WMCreatePLString(_("Styles")),
		WMCreatePLString("OPEN_MENU"),
		WMCreatePLString(style_paths),
		NULL
	);
	WMAddToPLArray(L1Menu, L2Menu);

	/* Root -> Appearance -> Icon Sets */
	L2Menu = WMCreatePLArray(
		WMCreatePLString(_("Icon Sets")),
		WMCreatePLString("OPEN_MENU"),
		WMCreatePLString(icon_paths),
		NULL
	);
	WMAddToPLArray(L1Menu, L2Menu);

	/* Root -> Appearance -> Background */
	L2Menu = WMCreatePLArray(WMCreatePLString(_("Background")), NULL);

	/* Root -> Appearance -> Background -> Solid */
	L3Menu = WMCreatePLArray(WMCreatePLString(_("Solid")), NULL);

#define SOLID_BACK(label, colorspec)									\
	L4Menu = WMCreatePLArray(									\
		WMCreatePLString(label),								\
		WMCreatePLString("EXEC"),								\
		WMCreatePLString("wdwrite WindowMaker WorkspaceBack '(solid, \"" colorspec "\")'"),	\
		NULL											\
	);												\
	WMAddToPLArray(L3Menu, L4Menu)

	/* Root -> Appearance -> Background -> Solid -> <color> */
	SOLID_BACK(_("Black"), "black");
	SOLID_BACK(_("Blue"), "#505075");
	SOLID_BACK(_("Indigo"), "#243e6c");
	SOLID_BACK(_("Bluemarine"), "#243e6c");
	SOLID_BACK(_("Purple"), "#554466");
	SOLID_BACK(_("Wheat"), "wheat4");
	SOLID_BACK(_("Dark Gray"), "#333340");
	SOLID_BACK(_("Wine"), "#400020");
#undef SOLID_BACK
	WMAddToPLArray(L2Menu, L3Menu);

	/* Root -> Appearance -> Background -> Gradient */
	L3Menu = WMCreatePLArray(WMCreatePLString(_("Gradient")), NULL);

#define GRADIENT_BACK(label, fcolorspec, tcolorspec)							\
	L4Menu = WMCreatePLArray(									\
		WMCreatePLString(label),								\
		WMCreatePLString("EXEC"),								\
		WMCreatePLString("wdwrite WindowMaker WorkspaceBack '(vgradient, \"" 			\
		    fcolorspec "\", \"" tcolorspec "\"'"),						\
		NULL											\
	);												\
	WMAddToPLArray(L3Menu, L4Menu)

	/* Root -> Appearance -> Background -> Gradient -> <color> */
	L4Menu = WMCreatePLArray(
		WMCreatePLString(_("Sunset")),
		WMCreatePLString("EXEC"),
		WMCreatePLString("wdwrite WindowMaker WorkspaceBack "
		    "'(mvgradient, deepskyblue4, black, deepskyblue4, tomato4)'"),
		NULL
	);
	WMAddToPLArray(L3Menu, L4Menu);
	GRADIENT_BACK(_("Sky"), "blue4", "white");
	GRADIENT_BACK(_("Blue Shades"), "#7080a5", "#101020");
	GRADIENT_BACK(_("Indigo Shades"), "#746ebc", "#242e4c");
	GRADIENT_BACK(_("Purple Shades"), "#654c66", "#151426");
	GRADIENT_BACK(_("Wheat Shades"), "#a09060", "#302010");
	GRADIENT_BACK(_("Grey Shades"), "#636380", "#131318");
	GRADIENT_BACK(_("Wine Shades"), "#600040", "#180010");
#undef GRADIENT_BACK
	WMAddToPLArray(L2Menu, L3Menu);

	/* Root -> Appearance -> Background -> Images */
	L3Menu = WMCreatePLArray(
		WMCreatePLString(_("Images")),
		WMCreatePLString("OPEN_MENU"),
		WMCreatePLString("-noext $HOME/GNUstep/Library/WindowMaker/Backgrounds WITH wmsetbg -u -t"),
		NULL
	);
	WMAddToPLArray(L2Menu, L3Menu);

	WMAddToPLArray(L1Menu, L2Menu);

	/* Root -> Appearance -> Save Theme */
	L2Menu = WMCreatePLArray(
		WMCreatePLString(_("Save Theme")),
		WMCreatePLString("SHEXEC"),
		WMCreatePLString("getstyle -p \"%a(Theme name, Name to save theme as)\""),
		NULL
	);
	WMAddToPLArray(L1Menu, L2Menu);

	/* Root -> Appearance -> Save IconSet */
	L2Menu = WMCreatePLArray(
		WMCreatePLString(_("Save IconSet")),
		WMCreatePLString("SHEXEC"),
		WMCreatePLString("geticonset $HOME/GNUstep/Library/WindowMaker/IconSets/"
			"\"%a(IconSet name,Name to save icon set as)\""),
		NULL
	);
	WMAddToPLArray(L1Menu, L2Menu);
	WMAddToPLArray(RMenu, L1Menu);

	/* Root -> Workspaces */
	L1Menu = WMCreatePLArray(
		WMCreatePLString(_("Workspaces")),
		WMCreatePLString("WORKSPACE_MENU"),
		NULL
	);
	WMAddToPLArray(RMenu, L1Menu);

	/* Root -> Workspace */
	L1Menu = WMCreatePLArray(WMCreatePLString(_("Workspace")), NULL);
	L2Menu = WMCreatePLArray(
		WMCreatePLString(_("Hide Others")),
		WMCreatePLString("HIDE_OTHERS"),
		NULL
	);
	WMAddToPLArray(L1Menu, L2Menu);

	/* Root -> Workspace -> Show All */
	L2Menu = WMCreatePLArray(
		WMCreatePLString(_("Show All")),
		WMCreatePLString("SHOW_ALL"),
		NULL
	);
	WMAddToPLArray(L1Menu, L2Menu);

	/* Root -> Workspace -> Arrange Icons */
	L2Menu = WMCreatePLArray(
		WMCreatePLString(_("Arrange Icons")),
		WMCreatePLString("ARRANGE_ICONS"),
		NULL
	);
	WMAddToPLArray(L1Menu, L2Menu);

	/* Root -> Workspace -> Refresh */
	L2Menu = WMCreatePLArray(
		WMCreatePLString(_("Refresh")),
		WMCreatePLString("REFRESH"),
		NULL
	);
	WMAddToPLArray(L1Menu, L2Menu);

	/* Root -> Workspace -> Save Session */
	L2Menu = WMCreatePLArray(
		WMCreatePLString(_("Save Session")),
		WMCreatePLString("SAVE_SESSION"),
		NULL
	);
	WMAddToPLArray(L1Menu, L2Menu);

	/* Root -> Workspace -> Clear Session */
	L2Menu = WMCreatePLArray(
		WMCreatePLString(_("Clear Session")),
		WMCreatePLString("CLEAR_SESSION"),
		NULL
	);
	WMAddToPLArray(L1Menu, L2Menu);
	WMAddToPLArray(RMenu, L1Menu);

	/* Root -> Configure Window Maker */
	L1Menu = WMCreatePLArray(
		WMCreatePLString(_("Configure Window Maker")),
		WMCreatePLString("EXEC"),
		WMCreatePLString("WPrefs"),
		NULL
	);
	WMAddToPLArray(RMenu, L1Menu);

	/* Root -> Info Panel */
	L1Menu = WMCreatePLArray(
		WMCreatePLString(_("Info Panel")),
		WMCreatePLString("INFO_PANEL"),
		NULL
	);
	WMAddToPLArray(RMenu, L1Menu);

	/* Root -> Restart Window Maker */
	L1Menu = WMCreatePLArray(
		WMCreatePLString(_("Restart Window Maker")),
		WMCreatePLString("RESTART"),
		NULL
	);
	WMAddToPLArray(RMenu, L1Menu);

	/* Root -> Other Window Managers [-> <other window manager> ...] */
	other_window_managers();

	/* Root -> Lock Screen */
	t = wfindfile(path, "xlock");
	if (t) {
		L1Menu = WMCreatePLArray(
			WMCreatePLString(_("Lock Screen")),
			WMCreatePLString("EXEC"),
			WMCreatePLString("xlock -allowroot -usefirst -mode matrix"),
			NULL
		);
		WMAddToPLArray(RMenu, L1Menu);
		wfree(t);
	}

	/* Root -> Exit Window Maker */
	L1Menu = WMCreatePLArray(
		WMCreatePLString(_("Exit Window Maker")),
		WMCreatePLString("EXIT"),
		NULL
	);
	WMAddToPLArray(RMenu, L1Menu);

	printf("%s", WMGetPropListDescription(RMenu, True));
	puts("");

	return 0;
}

/*
 * Creates an L2Menu made of L3Menu items
 * Attaches to L1Menu
 * - make sure previous menus of these levels are
 *   attached to their parent before calling
 */
static void find_and_write(const char *group, char *list[][2], int this_is_terminals)
{
	int i, argc;
	char *t, **argv, buf[PATH_MAX];

	/* or else pre-existing menus of these levels
	 * will badly disturb empty group detection */
	L2Menu = NULL;
	L3Menu = NULL;

	i = 0;
	while (list[i][0]) {
		/* Before checking if app exists, split its options */
		wtokensplit(list[i][1], &argv, &argc);
		t = wfindfile(path, argv[0]);
		if (t) {
			/* find a terminal to be used for cmnds that need a terminal */
			if (this_is_terminals && !terminal)
				terminal = wstrdup(list[i][1]);
			if (*(argv[argc-1]) != '!') {
				L3Menu = WMCreatePLArray(
					WMCreatePLString(_(list[i][0])),
					WMCreatePLString("EXEC"),
					WMCreatePLString(list[i][1]),
					NULL
				);
			} else {
				char comm[PATH_MAX], *ptr;

				strncpy(comm, list[i][1], sizeof(comm) - 1);
				comm[sizeof(comm) - 1] = '\0';

				/* delete character " !" from the command */
				ptr = strchr(comm, '!');
				if (ptr != NULL) {
					while (ptr > comm) {
						if (!isspace(ptr[-1]))
							break;
						ptr--;
					}
					*ptr = '\0';
				}
				snprintf(buf, sizeof(buf), "%s -e %s", terminal ? terminal : "xterm" , comm);

				/* Root -> Applications -> <category> -> <application> */
				L3Menu = WMCreatePLArray(
					WMCreatePLString(_(list[i][0])),
					WMCreatePLString("EXEC"),
					WMCreatePLString(buf),
					NULL
				);
			}
			if (!L2Menu)
				L2Menu = WMCreatePLArray(
					WMCreatePLString(group),
					NULL
				);
			WMAddToPLArray(L2Menu, L3Menu);
			wfree(t);
		}
		i++;
	}
	if (L2Menu)
		WMAddToPLArray(L1Menu, L2Menu);
}

/*
 * Creates an L1Menu made of L2Menu items
 * - make sure previous menus of these levels are
 *   attached to their parent before calling
 * Attaches to RMenu
 */
static void other_window_managers(void)
{
	int i;
	char *t, buf[PATH_MAX];

	/* or else pre-existing menus of these levels
	 * will badly disturb empty group detection */
	L1Menu = NULL;
	L2Menu = NULL;

	i = 0;
	while (other_wm[i][0]) {
		t = wfindfile(path, other_wm[i][1]);
		if (t) {
			snprintf(buf, sizeof(buf), _("Start %s"), _(other_wm[i][0]));
			/* Root -> Other Window Managers -> <other window manager> */
			L2Menu = WMCreatePLArray(
				WMCreatePLString(buf),
				WMCreatePLString("RESTART"),
				WMCreatePLString(other_wm[i][1]),
				NULL
			);
			if (!L1Menu)
				L1Menu = WMCreatePLArray(
					WMCreatePLString(_("Other Window Managers")),
					NULL
				);
			WMAddToPLArray(L1Menu, L2Menu);
			wfree(t);
		}
		i++;
	}
	if (L1Menu)
		WMAddToPLArray(RMenu, L1Menu);
}

noreturn void print_help(int print_usage, int exitval)
{
	printf("Usage: %s [-h] [-v]\n", __progname);
	if (print_usage) {
		puts("Writes a menu structure usable as ~/GNUstep/Defaults/WMRootMenu to stdout");
		puts("");
		puts("  -h, --help           display this help and exit");
		puts("  -v, --version        output version information and exit");
	}
	exit(exitval);
}
